### PR TITLE
feat(custom-resources): AwsResource - support custom resource type

### PR DESCRIPTION
**Pull request by netroy**
Allow specifying a custom resource type for `AwsResource`s.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts [hunk 1]
```diff
@@ -67,6 +67,10 @@ export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent
         region: call.region,
       });
 
+      if (event.RequestType === 'Delete' && call.physicalResourceIdPath) {
+        call.parameters[call.physicalResourceIdPath] = physicalResourceId;
+      }
```
Comment (jogold) [inline packages/@aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts]: If I understand it correctly, you want to be able to use the "current" physical id somewhere in you API call for the delete event?

The `physicalResourceIdPath` works the other way around: extracts the physical id from the response to return it to CloudFormation. You didn't update the documentation for this prop, so only you would know about this behavior here.

How about having some sort of tag (`{{PHYSICAL_ID}}`?) that could be used in the `parameters` prop and that would get replaced in the Lambda before the API call?

Can you maybe detail your use case?

@eladb opinion on this?

File: packages/@aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts [hunk 2]
```diff
@@ -172,6 +182,18 @@ export class AwsCustomResource extends cdk.Construct implements iam.IGrantable {
...
     }
 
+    let resourceType = props.resourceType;
+    if (resourceType) {
+      if (!new RegExp('^([a-zA-Z0-9_\\@\\-]+)$').test(resourceType)) {
+        throw new Error('`resourceType` can only contain alphanumeric characters, "_", "@", & "-"');
+      }
+      if (resourceType.length > 60) {
+        throw new Error('`resourceType` cannot be longer that 60 characters.');
+      }
```
Comment (jogold) [inline packages/@aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts]: This logic is already implemented here:
https://github.com/aws/aws-cdk/blob/590705586eb545c2c55cea1c0733b6d5f0876765/packages/%40aws-cdk/aws-cloudformation/lib/custom-resource.ts#L225-L241

File: packages/@aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts [hunk 3]
```diff
@@ -99,6 +99,16 @@ export interface AwsSdkCall {
 }
 
 export interface AwsCustomResourceProps {
+  /**
+   * The AWS Resource type.
+   * Valid values can include alphanumeric characters and the following characters: _@-
+   * If `resourceType` is specified, it will automatically be prefixed with `Custom::`
+   * Otherwise, it'll default to `AWS::CloudFormation::CustomResource`
+   *
+   * @default - AWS::CloudFormation::CustomResource
```
Comment (jogold) [inline packages/@aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts]: I would leave the default as `Custom::AWS`

File: packages/@aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts (hunk 1 above)
Comment (jogold) [inline packages/@aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts]: If I understand correctly you want to use the "current" physical id somewhere in your API call?

The `physicalResourceIdPath` works the other way around: it extracts the physical id from the response and returns it to CloudFormation. You haven't updated the documentation for this prop so only you would know about this behavior here.

How about having some sort of tag (`{{PHYSICAL_ID}}`?) that could be used in the `parameters` and would get replaced in the Lambda function before making the API call?

Can you detail your specific use case maybe?

@eladb opinion on this?

Comment (netroy) [reply packages/@aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts]: I'll add an example inline here, and will update the PR tomorrow morning.

When I create a `DatasetGroup` resource using `AwsCustomResource` like this:
```
new AwsCustomResource(this, 'DatasetGroup', {
  onCreate: {
    service: 'Personalize',
    action: 'createDatasetGroup',
    parameters: {
      name: 'TestDatasetGroup'
    },
    physicalResourceIdPath: 'datasetGroupArn'
  },
  onDelete: {
    service: 'Personalize',
    action: 'deleteDatasetGroup'
  }
})
```

The resource is created as expected.
But, when I tear down the stack, the call parameters for `deleteDatasetGroup` don't contain `datasetGroupArn`, the only parameter that's needed for the call. 

Same issues exists for other resources on Personalize & Forecast.

File: packages/@aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts (hunk 2 above)
Comment (netroy) [reply packages/@aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts]: I thought I saw that before 🤦‍♂ . Thanks for pointing it out.

File: packages/@aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts (hunk 1 above)
Comment (jogold) [reply packages/@aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts]: Wouldn't the following work here?
```ts
const groupName = 'TestDatasetGroup';
new AwsCustomResource(this, 'CreateDatasetGroup', {
  onCreate: {
    service: 'Personalize',
    action: 'createDatasetGroup',
    parameters: {
      name: groupName
    },
    physicalResourceIdPath: 'datasetGroupArn'
  },
  onDelete: {
    service: 'Personalize',
    action: 'deleteDatasetGroup',
    parameters: { // arn:aws:personalize:<region>:<account>>:dataset-group/TestDatasetGroup
      datasetGroupArn: cdk.Stack.of(this).formatArn({
        service: 'personalize',
        resource: 'dataset-group',
        resourceName: groupName
      })
    }
  }
});
```

Comment (netroy) [reply packages/@aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts]: Thanks @jogold, I'll try this out today.
If that works as expected, I can use that for now. 

It'd be good to know what @eladb prefers here, before I send out an update on the PR (or close it).

Cheers

Comment (eladb) [reply packages/@aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts]: I am with @jogold

**Pull request merged**